File: backend/infra/models/sqlalchemy/media.py
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base
from .mixins import CreatedAtMixin


class Media(CreatedAtMixin, Base):
    """Метаданные файла; само содержимое находится в S3-совместимом хранилище."""

    __tablename__ = "media"

    object_key: Mapped[str] = mapped_column(String(512), unique=True, index=True)
    bucket: Mapped[str] = mapped_column(String(255))
    original_filename: Mapped[str] = mapped_column(String(255))
    content_type: Mapped[str | None] = mapped_column(String(255), nullable=True)
    media_type: Mapped[str] = mapped_column(String(32))
    size: Mapped[int] = mapped_column(BigInteger())
    checksum: Mapped[str | None] = mapped_column(String(64), nullable=True, index=True)
    width: Mapped[int | None] = mapped_column(Integer(), nullable=True)
    height: Mapped[int | None] = mapped_column(Integer(), nullable=True)
    duration: Mapped[float | None] = mapped_column(nullable=True)
    uploaded_by: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: services/media/internal/domain/media.go
package domain

import (
	"time"
	"uuid"
)

// Media описывает метаданные объекта в S3-совместимом хранилище.
type Media struct {
	ID               uuid.UUID
	ObjectKey        string
	Bucket           string
	OriginalFilename string
	ContentType      string
	MediaType        string
	Size             int64
	Checksum         string
	Width            *int
	Height           *int
	Duration         *float64
	UploadedBy       uuid.UUID
	CreatedAt        time.Time
	DeletedAt        *time.Time
}

// CanBeManagedBy проверяет, может ли пользователь управлять метаданными файла.
func (m Media) CanBeManagedBy(userID uuid.UUID) bool {
	return m.UploadedBy == userID
}
